### DISCUSSION TITLE
What are some things I can do to improve simulation speed?

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#26115] [Q&A General] [answered]
I am running a simulation with 8 coupled variables, and it is incredibly slow. Compared to a similar setup which only has two variables (same equations as the former, but the former uses a mixed formulation hence each component of the flux is an independent variable), the time difference is staggering.

With the current mesh and temporal refinement the standard formulation can run ~400 timesteps in an hour or two, but the mixed formulation has been running for an hour and only completed 5 timesteps. Both are being run on the same HPC and are run in parallel with 25 CPUs per simulation with 4GB of RAM per CPU.

I wanted to know what suggestions you might have that could potentially reduce simulation time. The first thing I thought of was to hand-code the Jacobians (I'm currently using AD). Is there a tutorial that explains the MOOSE syntax for doing so in detail with an example? I'm having trouble finding one.

Also, this is the executioner block of the mixed formulation:

```
[Executioner]
  type = Transient
  solve_type = 'PJFNK'

  nl_rel_tol = 1e-8
  nl_abs_tol = 1e-8
  l_tol = 1e-5
  l_max_its = 300
  nl_max_its = 20

  dtmin = ${dt_val_min}
  dtmax= ${dt_val}
  
  start_time = ${start_val}
  end_time = ${t_val}
   
  [TimeStepper]
    type = IterationAdaptiveDT
    optimal_iterations = 15
    iteration_window = 3
    linear_iteration_ratio = 100
    growth_factor=1.5
    cutback_factor=0.5
    dt = ${dt_val}
  []
  [Predictor]
    type = SimplePredictor
    scale = 1.0
    skip_after_failed_timestep = true
  []
[]
```

Is PJFNK or NEWTON a better choice in this situation? I thought that since I didn't hand-code the Jacobians, that NEWTON wouldn't work. But while searching I saw an older post suggesting to use NEWTON if using AD. I guess I'm not fully sure what AD does/how it works.

The details of the problem I'm attempting to solve are here: https://github.com/idaholab/moose/discussions/26021
I was also wondering what petsc options might help and where I can learn about them.

I also wanted to ask about MPI- it seems that in this case the parallelization is not working as intended, as I'm getting this while it's solving:

![moose1](https://github.com/idaholab/moose/assets/52802063/7db362d9-fee9-436e-a29b-5cf0e15c0f36)

Which seems like it's using a lot of memory and/or splitting it is actually making it slower. It is taking way much more time to run than on my local computer (which is also very, very slow, but not as slow as this) which obviously has way less computational resources.

**Comment by GiudGiud** [comment, marked as answer]
Hello

If you use AD it s better to use Newton s method. Convergence rate is usually faster. I don't think there is a guarantee that this is true for all cases. 

The first thing to do here is to print the nonlinear and linear residuals and see which iteration is taking longer to converge now that there are 8 variables instead of 2. 

Then I would look at automatic scaling to see if it improves convergence. 

Then I would redo the preconditioning using a field split. With 8 variables you have to think which ones can be grouped and which ones can solved nearly on their own

Guillaume

**Reply by GiudGiud** [reply]
In fact you should print the residual on a per equation basis using the show_var_residual_norms in the Debug block to learn more

**Reply by lindsayad** [reply]
NEWTON is appropriate when your Jacobian is perfect, and the idea behind AD is that it makes the Jacobian perfect for you with the con that it is more expensive to run AD than if all your Jacobians were hand-coded.

The combination of AD and PJFNK can result in slow simulations because even though we circumvent most of the AD machinery when doing residual evaluations, we can't circumvent all of it. So you are getting some AD machinery for *every single* linear (and nonlinear) residual evaluation with PJFNK. With NEWTON, there are no residual function evaluations during the linear solve so you don't incur that cost.

If one is somehow forced to use PJFNK (which you shouldn't be) with AD, then you should try to use a powerful preconditioner that limits the number of linear residual evaluations. You are using the default parallel preconditioner which is block Jacobi with ILU on the sub-blocks; this is a weak preconditioner that will lead to lots of linear residual evaluations.

**Reply by lindsayad** [reply]
TLDR ... if you are using AD, definitely use NEWTON

**Reply by richmondodufisan** [reply]
I see. Thanks for the explanation. I switched to NEWTON and reduced the mesh refinement a bit so it would run faster, and here is what I got:

```
Time Step 1, time = 2.77778e-09, dt = 2.77778e-09


    Computing Initial Residual
      Finished Computing Residual                                                        [  5.15 s] [ 1248 MB]

  Skipping predictor this step

    Finished Computing Initial Residual                                                  [  5.15 s] [ 1248 MB]
    Finished Computing Residual                                                          [  5.03 s] [ 1248 MB]

    |residual|_2 of individual variables:
                 q_samp_x:   4.48827e-16
                 q_samp_y:   4.51781e-16
                 q_samp_z:   4.55923e-16
                 q_trans_x:  5.69263e-18
                 q_trans_y:  5.80489e-18
                 q_trans_z:  1.61914e-16
                 temp_samp:  1.11441e-17
                 temp_trans: 5.31787e-08

 0 Nonlinear |R| = 5.317868e-08

    Computing Jacobian.....                                                              [ 31.57 s] [ 1602 MB]

      0 Linear |R| = 5.317868e-08
      1 Linear |R| = 3.338002e-08
      2 Linear |R| = 1.965342e-08
      3 Linear |R| = 6.880915e-09
      4 Linear |R| = 6.663254e-09
      5 Linear |R| = 4.066381e-09
      6 Linear |R| = 4.056322e-09
      ...
    289 Linear |R| = 2.624816e-12
    290 Linear |R| = 2.624813e-12
    291 Linear |R| = 2.616900e-12
    292 Linear |R| = 2.600068e-12
    293 Linear |R| = 2.592508e-12
    294 Linear |R| = 2.591624e-12
    295 Linear |R| = 2.576122e-12
    296 Linear |R| = 2.565855e-12
    297 Linear |R| = 2.565026e-12
    298 Linear |R| = 2.529349e-12
    299 Linear |R| = 2.514333e-12
    300 Linear |R| = 2.504718e-12
  Linear solve did not converge due to DIVERGED_ITS iterations 300

    Finished Computing Residual                                                          [  5.34 s] [ 1920 MB]

    |residual|_2 of individual variables:
                 q_samp_x:   1.16411e-14
                 q_samp_y:   1.15637e-14
                 q_samp_z:   2.67838e-14
                 q_trans_x:  1.88732e-14
                 q_trans_y:  1.72999e-14
                 q_trans_z:  1.44232e-13
                 temp_samp:  3.04183e-14
                 temp_trans: 2.50005e-12

 1 Nonlinear |R| = 2.504720e-12

 Solve Converged!

  Finished Solving                                                                       [ 80.28 s] [ 1603 MB]


Outlier Variable Residual Norms:
  temp_trans: 2.500051e-12

Postprocessor Values:
+----------------+----------------+----------------+
| time           | integral_samp  | integral_trans |
+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
|   2.777778e-09 |   9.846154e-05 |   4.840187e-04 |
+----------------+----------------+----------------+
```

I'm guessing the temp_trans value being different could potentially be due to the elements there being less than optimal so I expect that (it's a thin layer).

The linear residuals are coming out to be on the order of 1e-12, and even then it still isn't "converging". I thought that I set the linear tolerance to 1e-5 with the statement

`l_tol = 1e-5`

in the executioner, but that clearly isn't being applied somehow. How can I fix that?

Also, what are the linear and nonlinear residuals even? By my understanding the equation system setup is of the format Ax = b where A is the Jacobian, and the residuals are calculated by Ax(guess) - b. So the closer that is to zero, the closer x is to the solution. And using NEWTON is the method by which the residuals are approximated. Where do the linear/nonlinear parts come in? And should the linear tolerance be greater or less than the nonlinear tolerance?

**Reply by lindsayad** [reply]
- The linear tolerance should be less than the nonlinear tolerance
- Newton's method is a way to linearize a nonlinear problem. Both NEWTON and PJFNK are different implementations of Newton's method. With NEWTON, the A matrix is explicitly formed. With PJFNK, the action of A on vectors (which is all that is needed for a Krylov method) is approximated using finite differences. This is why residual evaluations are performed during the linear solve: for computing those finite differences
- Both PJFNK and NEWTON require preconditioning during the linear solve. The stronger your preconditioner, the fewer linear iterations you will take. But also the stronger your preconditioner, generally the more costly each linear iteration will be. In your case, it looks like the default preconditioner is totally ineffective. How many degrees of freedom are in your reduced problem? Is this 3D or 2D or 1D?

**Reply by richmondodufisan** [reply]
Just to confirm, do you mean the linear tolerance should literally be a smaller number, or that it should be less restrictive (and so higher)?

Also figured out through trial and error that the relative linear tolerance is probably a better way to check for convergence than an absolute one (which I found to be set by `l_abs_tol`), at least for this problem.

It's a 3D problem. Even with a coarser mesh it's still relatively large:

```
Nonlinear System:
  Num DOFs:                240607
  Num Local DOFs:          63148
  Variables:               { "q_samp_x" "q_samp_y" "q_samp_z" } { "q_trans_x" "q_trans_y" "q_trans_z"
                             } "temp_samp" "temp_trans"
  Finite Element Types:    "LAGRANGE" "LAGRANGE" "LAGRANGE" "LAGRANGE"
  Approximation Orders:    "SECOND" "SECOND" "FIRST" "FIRST"

Auxiliary System:
  Num DOFs:                223544
  Num Local DOFs:          56132
  Variables:               { "avg_surf_temp" "sample_avg_surf_temp" } "bulk_gb_dist"
  Finite Element Types:    "LAGRANGE" "MONOMIAL"
  Approximation Orders:    "FIRST" "FIRST"
```

**Reply by lindsayad** [reply]
> Just to confirm, do you mean the linear tolerance should literally be a smaller number, or that it should be less restrictive (and so higher)?

Good clarification 😄 The linear tolerance should be looser than the nonlinear tolerance; it should be a *larger* number than for the nonlinear

> Also figured out through trial and error that the relative linear tolerance is probably a better way to check for convergence than an absolute one (which I found to be set by l_abs_tol), at least for this problem.

Yea I'm not sure I've every employed an absolute linear tolerance

> It's a 3D problem. Even with a coarser mesh it's still relatively large

Hmm yea this is pretty big. So for 3D calculations with 8 variables, the Jacobian evaluation done by AD probably will be pretty expensive because the sparsity pattern is large. For your linear preconditioning ... you could *try* LU and see what happens but the dense sparsity pattern and largish system size will likely be a problem. Exploring field split as @GiudGiud suggested is probably a good idea if LU is too expensive.

**Reply by richmondodufisan** [reply]
Great thanks!! I'll go tinker around with it

**Reply by lindsayad** [reply]
The matrix structure you get from your mixed formulation is identical to what you get in Navier-Stokes. You may want to look at some field split examples in that module in that vein. There is additional work on field split for navier-stokes in https://github.com/idaholab/moose/pull/24883 as well

**Reply by richmondodufisan** [reply]
Thanks, will look into that! I also got significant improvement in simulation speed by switching to newton and using automatic scaling. I will definitely be checking out the field split and also seeing if I can reduce to a 2D model. Thanks!!

**Reply by lindsayad** [reply]
oh that's good news!
